Allows guest users to access ggr

adrichem opened the pull request:
As discussed in https://github.com/aerokube/ggr/pull/104 

This runs in addition to Basic Authentication. Administrator can select which quota file to use for guests. All guest session share the same quota.

vania-pooh commented on proxy.go:
```diff
@@ -145,7 +145,7 @@ func serial() uint64 {
 
 func info(r *http.Request) (user, remote string) {
-	user = "unknown"
+	user = guestUserName
```
I think this should be `guestUserName` only when this feature is enabled.

vania-pooh commented on proxy.go:
```diff
@@ -211,7 +211,12 @@ func route(w http.ResponseWriter, r *http.Request) {
 	confLock.RUnlock()
-	if len(hosts) == 0 {
+
+	if(len(hosts) ==0 && user == guestUserName ) {
```
Please run `go fmt` on the entire project.

vania-pooh commented on proxy.go:
```diff
@@ -211,7 +211,12 @@ func route(w http.ResponseWriter, r *http.Request) {
+		reply(w, errMsg("Guest access is unavailable."), http.StatusUnauthorized)
+		log.Printf("[%d] [UNSUPPORTED_BROWSER] [%s] [%s] [%s]\n", id, user, remote, fmtBrowser(browser, version))
+		return
+	} else if len(hosts) == 0 {
```
I think we should return 401 in `WithSuitableAuthentication` instead. You can check whether guest quota exists like the following:
```
confLock.RLock()
if _, guestQuotaExists := quota[user]; !guestQuotaExists {
    //Return 401
}
confLock.RUnlock()
```

vania-pooh commented on proxy_test.go:
```diff
@@ -69,10 +69,13 @@ func hostportnum(u string) (string, int) {
 	listen = hostport(srv.URL)
 }
+func init() {
+	doInit()
```
Don't get this magic with `doInit`. Why we need to launch this initialization only is some tests? Seems to complicate things.

vania-pooh commented on proxy.go:
```diff
@@ -413,25 +413,32 @@ func requireBasicAuth(authenticator *auth.BasicAuth, handler func(http.ResponseW
+				} else {
+					handler(w, r)
+				}
+				confLock.RUnlock()
```
Sorry for changing this one more time but let's do like the following:
```
confLock.RLock()
defer confLock.RUnlock()
if _, ok := quota[guestUserName]; !ok {
```

adrichem replied:
Are you sure about this review issue?  The docs say: 
> If a goroutine holds a RWMutex for reading, it must not expect this or any other goroutine to be able to also take the read lock until the first read lock is released

vania-pooh replied:
@aandryashin any comments?

vania-pooh replied:
@adrichem after some discussion, please leave locking as is. Only change `isGuestUserPresent` -> `ok`.

vania-pooh commented on proxy.go:
```diff
@@ -413,25 +413,32 @@ func requireBasicAuth(authenticator *auth.BasicAuth, handler func(http.ResponseW
+				if _, guestQuotaExists := quota[guestUserName]; !guestQuotaExists {
+					//Return 401
+					reply(w, errMsg("Guest access is unavailable."), http.StatusUnauthorized)
+
```
Useless line.

vania-pooh commented on proxy.go:
```diff
@@ -413,25 +413,32 @@ func requireBasicAuth(authenticator *auth.BasicAuth, handler func(http.ResponseW
-				handler(w,r)
+				confLock.RLock()
+				if _, guestQuotaExists := quota[guestUserName]; !guestQuotaExists {
+					//Return 401
```
Useless comment.

aandryashin commented:
We should use

defer confLock.RUnlock()


2017-07-28 14:09 GMT+03:00 Ivan Krutov <notifications@github.com>:

> *@vania-pooh* commented on this pull request.
> ------------------------------
>
> In proxy.go
> <https://github.com/aerokube/ggr/pull/107#discussion_r130065995>:
>
> >  				//Run the handler as unauthenticated user
> -				handler(w,r)
> +				confLock.RLock()
> +				if _, guestQuotaExists := quota[guestUserName]; !guestQuotaExists {
> +					//Return 401
> +					reply(w, errMsg("Guest access is unavailable."), http.StatusUnauthorized)
> +
> +				} else {
> +					handler(w, r)
> +				}
> +				confLock.RUnlock()
>
> @aandryashin <https://github.com/aandryashin> any comments?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aerokube/ggr/pull/107#discussion_r130065995>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABWNbekxshDbe28x6Wm9NqgG577m60Ywks5sScFXgaJpZM4Olc3t>
> .
>

adrichem commented:
Ok, I squashed the commits and removed the defer on the RUnlock() after noticing that comment https://github.com/aerokube/ggr/pull/107#discussion_r130077245 is later than https://github.com/aerokube/ggr/pull/107#issuecomment-318629799 

Should be all complete now. Thanks for the input :)

vania-pooh commented on proxy.go:
```diff
@@ -408,6 +413,37 @@ func requireBasicAuth(authenticator *auth.BasicAuth, handler func(http.ResponseW
+				}
+			} else {
+				//Check that authentication is correct
+				userName := authenticator.CheckAuth(r)
```
Take a look at `authenticator.Wrap` method implementation used inside `requireBasicAuth`. It seems that this branch is a duplicate of this code. In that case the entire method can be simplified to something like:
```
func WithSuitableAuthentication(authenticator *auth.BasicAuth, handler func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
        return func(w http.ResponseWriter, r *http.Request) {
                _, _, basicAuthPresent := r.BasicAuth()
                if !basicAuthPresent && guestAccessAllowed {
                        // Guest authentication logic as currently implemented
                        // ...
                } else {
                        requireBasicAuth(authenticator, handler)(w, r)
                }
        }
}
```

adrichem replied:
Good idea. I didn't copy/paste your suggestion as-is, because I want to  avoid the call to `r.BasicAuth()` when guests are not allowed. 
```
if !guestAccessAllowed {
    //All requests require authentication
    ...
} else if _, _, basicAuthPresent := r.BasicAuth(); !basicAuthPresent {
    //Run the handler as unauthenticated user
    ...
} else {
    //Run the handler using basic authentication
    ...
}
```

vania-pooh replied:
Got it.

vania-pooh commented:
@adrichem thank you for contribution! You can try these changes with `aerokube/ggr:latest`.

vania-pooh commented:
@adrichem this functionality was released in version 1.2.6.
